All-reduce Cuda E2E Python test

sogartar opened the pull request:
Add an E2E test with StableHLO input that test all-reduce on 1 and 2 devices.

Here is assumed that at some point there will be a CI test runner with 2 Nvidia GPUs.

Depends on #14384.

sogartar committed:
When Initializing MPI don't fail if it is already initialized

It may be the case that the caller may have already initialized MPI in
this process.

sogartar committed:
In CI allow logical AND CTest label selection

Allows using "--label-regex regex1 --label-regex regex2".
Run tests where each regular expression matches at least one label.

sogartar committed:
Add CUDA collective all_reduce E2E Python test

ScottTodd commented on .github/workflows/ci.yml:
```diff
@@ -282,10 +282,15 @@ jobs:
+            iree-ctest-label-regex:
+              --label-regex ^requires-gpu|^driver=vulkan$|^driver=cuda$
+            iree-nvidia-sm80-tests-disable: 1
+          - runner-type: 2-gpus
```
> Here is assumed that at some point there will be a CI test runner with 2 Nvidia GPUs.

Can we add such a runner? Do we have a plan for that already? @pzread 
I think this would fail the CI job until such a runner is added. (We can approve the workflow runs to find out)

If adding a runner would be difficult (technically or on budgets), some options are:
* adding the test but not having it run on CI
* adding a separate CI job that runs nightly / on a schedule instead of on presubmit

pzread replied:
We haven't had a solid plan to add 2 GPU VM (I assume it will be a2-highgpu-2g). Although there are some requests from benchmarking.

 For the cost, I'd like to only have 1 VM in postsubmit at the beginning (its cost = 2x 1GPU VM). If there is a reasonable reason to run 2 GPUs tests in CI, I'm fine to add it). From the PR description, I'm not quite clear about why these tests need to run on 2GPUs, it will be great to give out more details to help us decide (like what do these tests cover in addition to single GPU tests).

sogartar replied:
In NCCL you can't put more than one rank per GPU. I will add explanation in the test.

ScottTodd replied:
To make progress on this PR, I'd suggest adding the test without a CI job that runs it. We can add a postsubmit CI job with multiple GPUs separately, but that work can happen independently.

(Seeing the proposed CI changes does help with the review, at least, but I think some of them can be reverted here / moved to a separate PR)

sogartar replied:
I removed the 2-GPU test from the CI job and the installation of OpenMPI from the Docker files.

ScottTodd commented on .github/workflows/ci.yml:
```diff
@@ -282,10 +282,15 @@ jobs:
+            iree-nvidia-sm80-tests-disable: 1
+          - runner-type: 2-gpus
+            iree-ctest-label-regex:
+              --label-regex ^driver=cuda$ --label-regex ^device-count=2$
```
This argument name could be changed to `iree-ctest-extra-args` since `--label-regex` is inlined. Or, the regex values used here could be merged into a single regex (is that possible here?) and the old functionality could be maintained.

sogartar replied:
I renamed it.

ScottTodd commented on build_tools/docker/dockerfiles/nvidia.Dockerfile:
```diff
@@ -59,3 +59,6 @@ RUN wget https://developer.download.nvidia.com/compute/cuda/12.1.1/local_install
 ENV PATH=${PATH}:/usr/local/cuda/bin/
+
+# Collectives: NCCL+OpenMPL
+RUN sudo apt-get -y install libnccl-dev=2.18.1-1+cuda12.1 libopenmpi-dev
```
Docker images will need to be rebuilt to pick up these changes.

sogartar replied:
I removed this.

ScottTodd commented on tests/e2e/collectives/run_shard.py:
This file could use a docstring at the top explaining its use.

sogartar replied:
I added some explanation. I renamed it `run_shard` -> `run_rank` and added documentation to the `run_ranks` function.

ScottTodd commented on tests/e2e/collectives/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+    LABELS
+      "requires-gpu-nvidia"
+      "driver=cuda"
+      "device-count=2"
```
How does this test behave on machines with only one GPU device? Having tests require specific hardware configurations has been a common point of confusion for developers (https://github.com/openxla/iree/issues/12042), so I'm wondering what options are available here to feature detect at configure/build/run-time, if it would make sense to add a CMake variable like `IREE_ENABLE_MULTIGPU_TESTS`, etc.

sogartar replied:
> How does this test behave on machines with only one GPU device?

If there are less than 2 devices it will fail during initialization with an error coming from NCCL.

We can have `IREE_ENABLE_MULTIGPU_TESTS`, but I thought that test labels are useful for just that case.

> I'm wondering what options are available here to feature detect at configure/build/run-time

The configuration whether to run a test should be pragmatically simple to avoid situations where a test that needs to run is by mistake disabled due to a programming error in the test configuration code.

ScottTodd replied:
The test labels do cover that case, but we'd still like for as many tests as possible to pass when just using `ctest` without any arguments (https://openxla.github.io/iree/building-from-source/getting-started/#running-tests)

It's probably fine here without adding an extra CMake variable, at least if the tests have a descriptive enough name that people can see the failures and understand reasonably quickly that they are expected to fail on their setups. On that note, how do you feel about putting "multi_gpu" in the test name, or something similar?

sogartar replied:
I renamed the tests to `collectives_test_single_gpu` and `collectives_test_multi_gpu`.

sogartar committed:
Address PR comments

ScottTodd commented on tests/e2e/collectives/collectives_test.py:
```diff
@@ -0,0 +1,170 @@
+import iree.runtime
+from iree.runtime.array_interop import DeviceArray
+import os
+from numpy.typing import ArrayLike
```
https://github.com/openxla/iree/actions/runs/5626159982/job/15295889995?pr=14418#step:7:232

```
ModuleNotFoundError: No module named 'numpy.typing'
```

Is that resolved by `mpi4py==3.1.4`? We might need to rebuild the Docker images to pull in any new python package deps.

sogartar replied:
I added `numpy` to the run requirements of the test.

ScottTodd replied:
Hmm... we already have one requirements file declare numpy: https://github.com/openxla/iree/blob/73b00a0559b07022fbef1d910e43ed8b674e83e8/runtime/bindings/python/iree/runtime/build_requirements.txt#L9

I can probably help untangle this, but it will take a few days (sorry for the delays... it's been a busy week x_x)

ScottTodd commented on .github/workflows/ci.yml:
```diff
@@ -282,10 +282,15 @@ jobs:
-            iree-ctest-label-regex: ^requires-gpu|^driver=vulkan$|^driver=cuda$
+            iree-ctest-extra-args:
+              --label-regex ^requires-gpu|^driver=vulkan$|^driver=cuda$
+            iree-nvidia-sm80-tests-disable: 1
```
Looks like this should also filter `device-count=2`, since the `test_gpu` job ran `iree/tests/e2e/collectives/collectives_test_multi_gpu`

https://github.com/openxla/iree/actions/runs/5626159982/job/15295889995?pr=14418#step:7:995

sogartar replied:
Done.

sogartar committed:
PR review round 2

Add missing run dependency to numpy.
Remove CI job requiring 2 GPUs.
Remove installation of MPI from Docker files.

sogartar committed:
Fix YAML formatting

ScottTodd commented on .github/workflows/ci.yml:
```diff
@@ -282,10 +282,12 @@ jobs:
+            iree-ctest-extra-args:
+              --label-regex ^requires-gpu|^driver=vulkan$|^driver=cuda$
+              --label-exclude ^device-count=2$
             iree-nvidia-sm80-tests-disable: 1
```
The latest CI failures suggest that nvidia-sm80 tests are running when they shouldn't be... https://github.com/openxla/iree/actions/runs/5670372477/job/15365761577?pr=14418#step:7:1160

`/work/runtime/src/iree/hal/drivers/cuda/native_executable.c:102: INTERNAL; CUDA driver error 'CUDA_ERROR_INVALID_PTX' (218): a PTX JIT compilation failed;`

```
The following tests FAILED:
	 11 - iree/tests/e2e/collectives/collectives_test_single_gpu (Failed)
	 12 - iree/tests/e2e/collectives/collectives_test_multi_gpu (Failed)
	 72 - iree/tests/e2e/matmul/e2e_matmul_direct_f32_gpu_large_LLVMGPUMatmulTensorCore_cuda_cuda (Failed)
	 73 - iree/tests/e2e/matmul/e2e_matmul_direct_f32_gpu_large_unaligned_cuda_cuda (Failed)
	 74 - iree/tests/e2e/matmul/e2e_matmul_direct_f16_gpu_large_unaligned_cuda_cuda (Failed)
	 75 - iree/tests/e2e/matmul/e2e_matmul_direct_f32_gpu_large_mma_sync_LLVMGPUMatmulTensorCoreMmaSync_cuda_cuda (Failed)
	 76 - iree/tests/e2e/matmul/e2e_matmul_direct_f16_gpu_large_LLVMGPUMatmulTensorCore_cuda_cuda (Failed)
	 77 - iree/tests/e2e/matmul/e2e_matmul_direct_f16_gpu_large_mma_sync_LLVMGPUMatmulTensorCoreMmaSync_cuda_cuda (Failed)
	734 - iree/tests/transform_dialect/cuda/double_mma_layout_analysis.mlir.test (Failed)
	735 - iree/tests/transform_dialect/cuda/mma_elemwise_layout_analysis.mlir.test (Failed)
	736 - iree/tests/transform_dialect/cuda/mma_reduction_layout_analysis.mlir.test (Failed)
	737 - iree/tests/transform_dialect/cuda/mma_using_layout_analysis.mlir.test (Failed)
```

Maybe some issue with newlines in the ctest command coming from this.

sogartar replied:
I added a new flag to exclude multi-device tests due to the behavior of multiple ctest `--label-exclude` parameters.

sogartar committed:
Disable multi-device test through a CI flag

sogartar committed:
Change back CI option iree-ctest-extra-args to iree-ctest-label-regex

ScottTodd reviewed (commented):
LGTM, but it looks like we'll need to update the Docker files, which might need some tweaks to which requirements files are used. That will be easiest for someone with permissions on the cloud project to do (probably me or @pzread)

https://github.com/openxla/iree/actions/runs/5675890178/job/15400747021?pr=14418#step:7:237

`ModuleNotFoundError: No module named 'numpy.typing'`

sogartar committed:
Remove usage of numpy.typing.ArrayLike

ScottTodd commented on tests/e2e/collectives/run_requirements.txt:
```diff
@@ -0,0 +1,2 @@
+mpi4py==3.1.4
```
On my Linux (Debian) machine:

```shell
python -m pip install -r tests/run_requirements.txt

...
      _configtest.c:2:10: fatal error: 'mpi.h' file not found
      #include <mpi.h>
               ^~~~~~~
      1 error generated.
      failure.
      removing: _configtest.c _configtest.o
      error: Cannot compile MPI programs. Check your configuration!!!
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for mpi4py
Failed to build mpi4py
ERROR: Could not build wheels for mpi4py, which is required to install pyproject.toml-based projects
```

Is that a hard requirement for the Python test? Does the runtime C code need similar dependencies?

Some ideas:

* We could use https://cmake.org/cmake/help/latest/module/FindMPI.html to enable/disable the test 
* https://openxla.github.io/iree/building-from-source/getting-started/#dependencies could mention `python -m pip install -r tests/run_requirements.txt` for running tests in addition to `python -m pip install -r runtime/bindings/python/iree/runtime/build_requirements.txt` for building the python bindings
  * I'd prefer for the run requirements to install cleanly without additional system package installs though
* We can jump through some hoops in our Docker images as needed (e.g. installing Open MPI or whatever else https://pypi.org/project/mpi4py/ wants), but the developer experience for the core project should work without a hard requirement to install specific requirements (e.g. CUDA, Vulkan SDK, etc.)
* Does pip support conditional requirements? If not, maybe we shouldn't put a direct include between `tests/run_requirements.txt` and `tests/e2e/collectives/run_requirements.txt` and just keep it optional. Not great either way :/

ScottTodd replied:
Maybe we could replace `rank = MPI.COMM_WORLD.Get_rank()` with some call into IREE's Python API, and have that use some underlying C code that doesn't require an extra package install?

sogartar replied:
I added a CMake flag to exclude collectives tests.

> * Does pip support conditional requirements? If not, maybe we shouldn't put a direct include between `tests/run_requirements.txt` and `tests/e2e/collectives/run_requirements.txt` and just keep it optional.

Pip requirements files seem to not have conditional control of what should be installed.
Conditional test run environment setup should be driven by the configuration of the build. It would probably need to generate the installation procedure.
Do you want me to include such facilities in this PR? Right now the pip requirements for the collectives tests are not included anywhere.

> Maybe we could replace rank = MPI.COMM_WORLD.Get_rank() with some call into IREE's Python API, and have that use some underlying C code that doesn't require an extra package install?

This approach will work during build time. In IREE the MPI library is dynamically loaded with `dlopen`.
During runtime you would still need `libmpi.so`. There is not pip package for that. At least OpenMPI does not have one.

It seems NCCL can be installed with pip
https://pypi.org/project/nvidia-nccl-cu11/
https://pypi.org/project/nvidia-nccl-cu12/

ScottTodd reviewed (commented):
Okay, I finally had some time to look at the dependency requirements on a local machine. Sorry for the delays - I really appreciate how patient you've been.

sogartar committed:
Add CMake flag to exclude collectives tests

Remove mpi4py requirements form CI.

ScottTodd reviewed (approved):
Looks good to land this as-is. There is a pytype warning but I don't really care about it and I think we can ignore it: https://github.com/openxla/iree/actions/runs/5716894928/job/15490623135?pr=14418#step:6:11

Follow-up changes can install deps in Docker, add a CI build with multiple GPUs, etc.
